hcl2template/types.variables.go

```diff
@@ -669,8 +669,11 @@ func (cfg *PackerConfig) collectInputVariableValues(env []string, files []*hcl.F
 					Detail: fmt.Sprintf("A %q variable was set but was "+
 						"not found in known variables. To declare "+
```

Comment:
I think part of the confusion, especially for users coming from JSON, is that Packer is reporting the variable as not found as a variable but it was indeed found Packer just doesn't know what to do with it. In JSON the var-files allowed to declare and set in a single file, which is the cause for the confusion with HCL. I think we can improve this error a bit be explicitly saying that the variable was set but not declared. 

```suggestion
						"not declared as an input variable. To declare "+
```

Reply:
I wasn't aware that JSON varfiles let you declare and set variables, is there a reason why HCL won't let us do this? Maybe there's something in the grammar we can improve to fix this problem as this is indeed confusing for users who're not yet familiar with this behaviour.
Also I think it'd be worth it to walk through the docs to improve them, given how prevalent the reporting on this is, I suspect the delineation between variable declaration and assignation may not be clear enough. Maybe there should be a learning guide too?
Note that my last paragraph assumes a lot of things related to the docs, I didn't take a look at them, maybe all I'm suggesting is already here.

Reply:
> Also I think it'd be worth it to walk through the docs to improve them, given how prevalent the reporting on this is, I suspect the delineation between variable declaration and assignation may not be clear enough.

Agreed. I added the following comment to the open PR you pinged me on. "Looking at the PR it appears that we should start be being a bit more explicit in the docs about variable loading. I see https://www.packer.io/docs/templates/hcl_templates/variables#assigning-values-to-build-variables mentions assigning variables but is immediately followed by variable validation which seems like it should come after. We can start by rearranging the documentation." 

I think this a follow up PR to this error message.

>  Maybe there should be a learning guide too?

Yes, this is a learn guide and a talk that we should do. Migrating from JSON to HCL.